src/control/lib/daos/container_property.go

```diff
@@ -0,0 +1,1214 @@
+	keys = make([]string, 0, len(cpl.entries))
+	for _, prop := range cpl.Properties() {
+		if prop.IsReadOnly() && excReadOnly {
+			continue
+		}
+
+		keys = append(keys, prop.Name)
+		if prop.hdlr.deprecatedNames != nil {
+			keys = append(keys, prop.hdlr.deprecatedNames...)
+		}
+	}
+	sort.Strings(keys)
+	return
+}
+
+// Properties returns a slice of container property entries.
+func (cpl *ContainerPropertyList) Properties() (props []*ContainerProperty) {
+	for i := range cpl.entries {
+		props = append(props, newContainerProperty(i, &cpl.entries[i]))
+	}
+	return
+}
+
+// newContainerProperty returns an initialized container property.
+func newContainerProperty(listIdx int, entry *C.struct_daos_prop_entry) *ContainerProperty {
+	propType := ContainerPropType(entry.dpe_type)
+	hdlr := propHdlrs[propType.String()]
+
+	return &ContainerProperty{
+		property: property{
+			idx:   C.int(listIdx),
+			entry: entry,
+		},
+		hdlr:        hdlr,
+		Type:        propType,
+		Name:        propType.String(),
+		Description: hdlr.shortDesc,
+	}
+}
+
+// IsReadOnly returns true if the property is read-only.
+func (cp *ContainerProperty) IsReadOnly() bool {
+	return cp.hdlr.readOnly
+}
+
+// Set accepts a string value and attempts to set the underlying property via
+// the property's handler. Returns an error if the property is read-only or
+// the value is invalid for the property.
+func (cp *ContainerProperty) Set(value string) error {
+	if cp.IsReadOnly() {
+		return errors.Errorf("property %q is read-only", cp.Name)
+	}
+
+	return cp.hdlr.execute(cp, value)
+}
+
+// StringValue returns a string representation of the property's value.
+func (cp *ContainerProperty) StringValue() string {
+	if cp.IsUnset() {
+		return "not set"
+	}
+	if cp.hdlr.toString == nil {
+		// Should be caught by unit test.
+		panic(fmt.Sprintf("%s: no toString function set", cp.Name))
+	}
+
+	return cp.hdlr.toString(cp)
+}
+
+// String returns a string representation of the property suitable for debug
+// logging.
+func (cp *ContainerProperty) String() string {
+	return fmt.Sprintf("%s: %s", cp.Name, cp.StringValue())
+}
+
+// SettableValues returns a slice of strings corresponding to the set of
+// predefined valid values for the property, if any.
+func (p *ContainerProperty) SettableValues() []string {
+	if p.hdlr == nil || p.hdlr.valHdlrs == nil {
+		return nil
+	}
+
+	return p.hdlr.valHdlrs.keys()
+}
+
+func (p *ContainerProperty) MarshalJSON() ([]byte, error) {
+	if p == nil || p.entry == nil || p.hdlr == nil {
+		return nil, errors.New("nil property")
+	}
+
+	// Normal case, just use the stringer.
+	jsonValue := func(p *ContainerProperty) any {
+		return p.hdlr.toString(p)
+	}
+	// Special-case situations for when the string representation
+	// of a value would be wrong for JSON consumers (e.g. human-readable
+	// numeric values that would have to be converted back to a number).
+	switch p.Type {
+	case ContainerPropHighestOid,
```

Comment:
might want to add in the notes to developers adding new properties above that they might need to update this list for specific cases

Reply:
Moved the MarshalJSON method up so that it's right under the property map. That's a better place for it because it's on the side where developers will look for things that need to be updated for new properties.